### PR TITLE
Leverage authResult from complete hpp payment call to cancel the payment

### PR DESCRIPTION
**Pull request by daliwei**
cc @pierre for review.

### CODE REVIEW

File: src/test/java/org/killbill/billing/plugin/adyen/TestRemoteBase.java [hunk 1]
```diff
@@ -89,13 +89,14 @@
     protected Properties properties;
     protected String merchantAccount;
 
-    @BeforeClass(groups = {"slow", "integration"})
-    public void setUpBeforeClassCommon() throws Exception {
+    @BeforeClass(groups = {"slow"})
```
Comment (pierre) [inline src/test/java/org/killbill/billing/plugin/adyen/TestRemoteBase.java]: It would be easier for the merge into `master` to take the latest fix instead, see https://github.com/killbill/killbill-adyen-plugin/commit/77f9a6811726069edf24c85aedbd28b95154d850#diff-9297c9fce44c5cf6a43ccaf9fae29ebc.

File: src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java [hunk 2]
```diff
@@ -238,15 +242,53 @@ public void testCancelExpiredPayPalPaymentNoNotification() throws Exception {
...
     @Test(groups = "integration")
     public void testCancelExpiredBoletoPayment() throws Exception {
         final Payment payment = triggerBuildFormDescriptor(ImmutableMap.<String, String>of(AdyenPaymentPluginApi.PROPERTY_CREATE_PENDING_PAYMENT, "true",
                                                                                            AdyenPaymentPluginApi.PROPERTY_AUTH_MODE, "true"),
                                                            TransactionType.AUTHORIZE);
         dao.updateResponse(payment.getTransactions().get(0).getId(), PaymentServiceProviderResult.PENDING, ImmutableList.<PluginProperty>of(new PluginProperty("paymentMethod", "boletobancario_santander", false)), context.getTenantId());
 
-        final Period expirationPeriod = adyenConfigProperties.getPendingPaymentExpirationPeriod("boletobancario_santander");
-        assertEquals(expirationPeriod.toString(), "P7D");
+        final Period expirationPeriod = adyenConfigProperties.getPendingHppPaymentWithoutCompletionExpirationPeriod();
```
Comment (pierre) [inline src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: Can you explain the change in expectations here?

It doesn't make sense to me: for Boleto bancário, the customer needs a few days to go to a local kiosk in person, and pay the paper voucher.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java [hunk 3]
```diff
@@ -100,9 +100,14 @@ public Map getDirectory(final String merchantAccount,
         return directoryClient.getDirectory(params);
     }
 
-    // Used to verify completion
-    @SuppressWarnings("deprecation")
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: Why this change? `Signer#signFormParameters` (`SHA1` version) is deprecated.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java [hunk 4]
```diff
@@ -100,9 +100,14 @@ public Map getDirectory(final String merchantAccount,
...
+
     public HppCompletedResult parseAndVerifyRequestIntegrity(final Map<String, String> requestParameterMap) {
+        return parseAndVerifyRequestIntegrity(requestParameterMap, adyenConfigProperties, logger);
+    }
+
+    public static HppCompletedResult parseAndVerifyRequestIntegrity(final Map<String, String> requestParameterMap,
+                                                                    final AdyenConfigProperties adyenConfigProperties,
+                                                                    final Logger logger) {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: Don't pass a `logger`, use the one from `AdyenPaymentServiceProviderHostedPaymentPagePort`.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java [hunk 5]
```diff
@@ -100,9 +100,14 @@ public Map getDirectory(final String merchantAccount,
...
 
-    // Used to verify completion
-    @SuppressWarnings("deprecation")
+
     public HppCompletedResult parseAndVerifyRequestIntegrity(final Map<String, String> requestParameterMap) {
+        return parseAndVerifyRequestIntegrity(requestParameterMap, adyenConfigProperties, logger);
+    }
+
+    public static HppCompletedResult parseAndVerifyRequestIntegrity(final Map<String, String> requestParameterMap,
+                                                                    final AdyenConfigProperties adyenConfigProperties,
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: Why do you need to pass these `properties`? They should be the same as the one from `AdyenPaymentServiceProviderHostedPaymentPagePort` (same tenant).

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java [hunk 6]
```diff
@@ -617,6 +622,59 @@ public T execute(final String merchantAccount, final PaymentData paymentData, fi
...
+        try {
+            PaymentApiWrapper paymentApiWrapper = new PaymentApiWrapper(killbillAPI, false);
+            paymentApiWrapper.transitionPendingTransaction(killbillAPI.getAccountUserApi().getAccountById(accountId, context),
+                                                           kbPaymentId,
+                                                           kbPaymentTransactionId,
+                                                           PaymentPluginStatus.CANCELED,
+                                                           context);
+        } catch (AccountApiException e) {
+            //Swallow exceptions so the original payment response record will still be updated.
+            logger.error(String.format("Unable to fetch the account based on accountId='{0}'", accountId));
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: No need for `String.format` with `SLF4j` (`{}` notation).

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java [hunk 7]
```diff
@@ -617,6 +622,59 @@ public T execute(final String merchantAccount, final PaymentData paymentData, fi
...
+                                  final CallContext context) {
+        try {
+            PaymentApiWrapper paymentApiWrapper = new PaymentApiWrapper(killbillAPI, false);
+            paymentApiWrapper.transitionPendingTransaction(killbillAPI.getAccountUserApi().getAccountById(accountId, context),
+                                                           kbPaymentId,
+                                                           kbPaymentTransactionId,
+                                                           PaymentPluginStatus.CANCELED,
+                                                           context);
+        } catch (AccountApiException e) {
+            //Swallow exceptions so the original payment response record will still be updated.
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Make sure to pass the full exception in the error message (`logger.error(format, arguments, e)`), so we understand the error.

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java [hunk 8]
```diff
@@ -106,6 +108,7 @@
 import com.google.common.collect.Maps;
 
 import static org.killbill.billing.plugin.adyen.api.mapping.UserDataMappingService.toUserData;
+import static org.killbill.billing.plugin.api.PluginProperties.merge;
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Nit: avoid static imports for this.

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java [hunk 9]
```diff
@@ -617,6 +622,59 @@ public T execute(final String merchantAccount, final PaymentData paymentData, fi
...
+                                                                          properties,
+                                                                          context.getTenantId());
+        }
+        return buildPaymentTransactionInfoPlugin(adyenResponsesRecord);
+    }
+
+    private boolean needCancelHppPayment(final PaymentServiceProviderResult authResult) {
+        //We only cancel the hpp payment during a complete request if the authResult is either canceled or refused.
+        //In all other cases, we keep the status in pending until the confirmation of a notification.
+        return ImmutableList.of(PaymentServiceProviderResult.CANCELLED,
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: This list (or set?) could be materialized (e.g. class constant), instead of rebuilding it all the time.

File: src/test/java/org/killbill/billing/plugin/adyen/TestRemoteBase.java (hunk 1 above)
Comment (daliwei) [reply src/test/java/org/killbill/billing/plugin/adyen/TestRemoteBase.java]: merged that commit to the PR branch.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java (hunk 3 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: Added back.

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java (hunk 4 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: Removed the use of static method. See b08b884

File: src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java (hunk 5 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/client/payment/service/AdyenPaymentServiceProviderHostedPaymentPagePort.java]: See b08b884. Removed the change that added this static method.

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java (hunk 6 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Updated.

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java (hunk 7 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Updated.

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java (hunk 8 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Updated

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java (hunk 9 above)
Comment (daliwei) [reply src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Updated

File: src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java (hunk 2 above)
Comment (daliwei) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: `getPendingHppPaymentWithoutCompletionExpirationPeriod` is used for those HPP payment methods where **no complete auth call** has been received.

To also leverage `paymentMethodToExpirationPeriod`, I expanded the method `getPendingHppPaymentWithoutCompletionExpirationPeriod` to check  `paymentMethodToExpirationPeriod` first. 

`PT50M` is just a hypothetical setting from the test properties file.

Comment (pierre) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: > `getPendingHppPaymentWithoutCompletionExpirationPeriod` is used for those HPP payment methods where **no complete auth call** has been received.

Yes, and this is what the test is verifying. The voucher has been created (`PENDING` transaction) and no completion call has been received yet (purchaser has yet to go to the store).

> To also leverage `paymentMethodToExpirationPeriod`, I expanded the method `getPendingHppPaymentWithoutCompletionExpirationPeriod` to check `paymentMethodToExpirationPeriod` first.
> 
> `PT50M` is just a hypothetical setting from the test properties file.

Understood, but the tests config being the same, how come the expiration period has changed? The expectations in `TestAdyenConfigProperties` for instance remain the same.

Sorry for being picky here but I want to make sure there is no subtle change in behavior around this (and if the test was badly setup in the first place, I'd rather fix it than expecting _bad_ / _unrealistic_ values).

Comment (daliwei) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: The expectation changes because the expiration of HPP payment method without a complete auth is set via the property of `pendingHppPaymentWithoutCompletionExpirationPeriod`. The change introduced the in the PR #93, but the test was not changed accordingly. 

The reason expectations in TestAdyenConfigProperties didn't change is because the properties are set manually instead of reading from the property file. Without a global `pendingHppPaymentWithoutCompletionExpirationPeriod` being explicitly set, it defaults back to default expiration date by payment method.

Comment (pierre) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: Are the tests passing for you?

I just pulled `daliwei/0.18.x-backport` as of a361e651e3ce33fa3b42d99175696e979d42b6af and I'm getting:

```
java.lang.AssertionError: expected [PT50M] but found [P7D]
Expected :PT50M
Actual   :P7D
```

Comment (daliwei) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: sorry, pinged the `adyen.properties` file to you offline.

Comment (pierre) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: Your test configuration has custom properties, hence the expectation changes:

```
org.killbill.billing.plugin.adyen.pendingPaymentExpirationPeriod=P7D
org.killbill.billing.plugin.adyen.pending3DsPaymentExpirationPeriod=PT1H
org.killbill.billing.plugin.adyen.pendingHppPaymentWithoutCompletionExpirationPeriod=PT50M
org.killbill.billing.plugin.adyen.sensitiveProperties=email|shippingAddress|purchaser_email|ip
```

Can you make sure the suites (`-Pmysql` and `-Pintegration-mysql`) pass with just the urls and credentials (rest should use the default config)?

```
org.killbill.billing.plugin.adyen.merchantAccount
org.killbill.billing.plugin.adyen.username
org.killbill.billing.plugin.adyen.password
org.killbill.billing.plugin.adyen.paymentUrl
org.killbill.billing.plugin.adyen.recurringUrl
org.killbill.billing.plugin.adyen.directoryUrl
org.killbill.billing.plugin.adyen.hmac.algorithm
org.killbill.billing.plugin.adyen.hmac.secret
org.killbill.billing.plugin.adyen.skin
org.killbill.billing.plugin.adyen.hpp.target
```

Comment (daliwei) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: @pierre Done in a34e15e

Comment (pierre) [reply src/test/java/org/killbill/billing/plugin/adyen/api/TestAdyenPaymentPluginApiHPP.java]: Thanks! 🎉
